### PR TITLE
[test] Allow setting the NUMA node

### PR DESCRIPTION
**Pull request by pb8o**
## Changes

Some changes in the tests framework. 

- Allow setting the NUMA node in integration tests.
- Extract the Network namespace logic out of the Jailer class, for more flexibility.
- Drop the resource path concept since it is not needed.

## Reason

This is not needed for any of the current tests, but it can be useful when building new performance tests.

## License Acceptance

By submitting this pull request, I confirm that my contribution is made under
the terms of the Apache 2.0 license. For more information on following
Developer Certificate of Origin and signing off your commits, please check
[`CONTRIBUTING.md`][3].

## PR Checklist

- [ ] If a specific issue led to this PR, this PR closes the issue.
- [ ] The description of changes is clear and encompassing.
- [ ] Any required documentation changes (code and docs) are included in this PR.
- [ ] API changes follow the [Runbook for Firecracker API changes][2].
- [ ] User-facing changes are mentioned in `CHANGELOG.md`.
- [ ] All added/changed functionality is tested.
- [ ] New `TODO`s link to an issue.
- [ ] Commits meet [contribution quality standards](https://github.com/firecracker-microvm/firecracker/blob/main/CONTRIBUTING.md#contribution-quality-standards).

---

- [x] This functionality cannot be added in [`rust-vmm`][1].

[1]: https://github.com/rust-vmm
[2]: https://github.com/firecracker-microvm/firecracker/blob/main/docs/api-change-runbook.md
[3]: https://github.com/firecracker-microvm/firecracker/blob/main/CONTRIBUTING.md

### CODE REVIEW

File: tests/integration_tests/functional/test_shut_down.py [hunk 1]
```diff
@@ -14,7 +14,6 @@ def test_reboot(test_microvm_with_api):
     Test reboot from guest.
     """
     vm = test_microvm_with_api
-    vm.jailer.daemonize = False
```
Comment (pb8o) [inline tests/integration_tests/functional/test_shut_down.py:17]: While developing this commit I found this was not needed by the test. I included it here, but I could also split it out as a separate commit if that helps.

File: tests/integration_tests/functional/test_serial_io.py [hunk 2]
```diff
@@ -145,7 +145,6 @@ def test_serial_dos(test_microvm_with_api):
     # Set up the microVM with 1 vCPU and a serial console.
     microvm.basic_config(
         vcpu_count=1,
-        add_root_device=False,
```
Comment (pb8o) [inline tests/integration_tests/functional/test_serial_io.py:148]: While developing this commit I found this was not needed by the test. I included it here, but I could also split it out as a separate commit if that helps.

File: tests/framework/microvm.py [hunk 3]
```diff
@@ -521,9 +525,17 @@ def spawn(
 
         # When the daemonize flag is on, we want to clone-exec into the
```
Comment (pb8o) [inline tests/framework/microvm.py:526]: Maybe I should move the comment to where it was before.

File: tests/integration_tests/security/test_jail.py [hunk 4]
```diff
@@ -107,6 +92,8 @@ def test_exec_file_not_exist(test_microvm_with_api, tmp_path):
 
     # Error case 1: No such file exists
     pseudo_exec_file_path = tmp_path / "pseudo_firecracker_exec_file"
+    fc_dir = Path("/srv/jailer") / pseudo_exec_file_path.name / test_microvm.id
```
Comment (roypat) [inline tests/integration_tests/security/test_jail.py:95]: nit: can probably make "/srv/jailer" a constant in defs.py

File: tests/framework/utils.py [hunk 5]
```diff
@@ -680,7 +680,7 @@ def populate_data_store(test_microvm, data_store):
 
 def start_screen_process(screen_log, session_name, binary_path, binary_params):
     """Start binary process into a screen session."""
-    start_cmd = "screen -L -Logfile {logfile} " "-dmS {session} {binary} {params}"
```
Comment (roypat) [inline tests/framework/utils.py:683]: Thank you! This weird space has been bothering me so much, but I kept forgetting to actually do something about it :joy:

**Review by roypat — DISMISSED** [review]

File: tests/framework/jailer.py [hunk 6]
```diff
@@ -44,10 +43,10 @@ def __init__(
         self,
         jailer_id,
         exec_file,
+        netns,
```
Comment (bchalios) [inline tests/framework/jailer.py:46]: Why is that not optional any more? Can't we have jailer environments without network namespace configured?

File: tests/framework/jailer.py [hunk 7]
```diff
@@ -103,7 +102,7 @@ def construct_param_list(self):
         if self.chroot_base is not None:
             jailer_param_list.extend(["--chroot-base-dir", str(self.chroot_base)])
         if self.netns is not None:
```
Comment (bchalios) [inline tests/framework/jailer.py:105]: Either we should make `netns` optional on the constructor or this if is not needed, right?

File: tests/framework/jailer.py [hunk 8]
```diff
@@ -68,7 +67,7 @@ def __init__(
         self.uid = uid
         self.gid = gid
         self.chroot_base = chroot_base
-        self.netns = netns if netns is not None else jailer_id
+        self.netns = netns
```
Comment (bchalios) [inline tests/framework/jailer.py:70]: If this always required now, should we check for it being not `None`?

File: tests/framework/microvm.py [hunk 9]
```diff
@@ -827,7 +828,7 @@ def ssh_iface(self, iface_idx=0):
         guest_ip = list(self.iface.values())[iface_idx]["iface"].guest_ip
         self.ssh_key = Path(self.ssh_key)
         return net_tools.SSHConnection(
-            netns=self.jailer.netns,
+            netns=self.netns.id,
```
Comment (bchalios) [inline tests/framework/microvm.py:841]: Is the change equivalent to the old one? Before we were assigning the `netns` object, now its `id`

File: tests/framework/microvm.py [hunk 10]
```diff
@@ -842,23 +843,30 @@ def ssh(self):
...
         self.base_path = Path(base_path)
         self.vms = []
         self.fc_binary_path = Path(fc_binary_path)
         self.jailer_binary_path = Path(jailer_binary_path)
+        self.kwargs = kwargs
 
     def build(self, kernel=None, rootfs=None, **kwargs):
         """Build a microvm"""
+        kwargs = self.kwargs | kwargs
+        microvm_id = kwargs.pop("microvm_id", str(uuid.uuid4()))
```
Comment (bchalios) [inline tests/framework/microvm.py:865]: what happens if `microvm_id` is not a key in `kwargs`?

File: tools/devctr/Dockerfile [hunk 11]
```diff
@@ -72,6 +72,7 @@ RUN apt-get update \
         musl-tools \
         # needed for integration tests
         net-tools iproute2 iperf3 socat fdisk \
+        numactl \
```
Comment (bchalios) [inline tools/devctr/Dockerfile:75]: Have you pushed a new container image?

File: tests/framework/jailer.py (hunk 6 above)
Comment (pb8o) [reply tests/framework/jailer.py:46]: hmm we can by passing `None`, but true. I don't remember why I did this. I will revert.

File: tests/framework/microvm.py (hunk 9 above)
Comment (pb8o) [reply tests/framework/microvm.py:841]: Yeah, before it was a string, which is the same as `id`

File: tests/framework/microvm.py (hunk 10 above)
Comment (pb8o) [reply tests/framework/microvm.py:865]: The 2nd parameter acts as default, `str(uuid.uuid4())`

File: tools/devctr/Dockerfile (hunk 11 above)
Comment (pb8o) [reply tools/devctr/Dockerfile:75]: No, I don't need this immediately, but if we are doing one soon it would be great to have this included.

File: tests/framework/jailer.py (hunk 6 above)
Comment (pb8o) [reply tests/framework/jailer.py:46]: Actually I am not sure how well no network namespace is supported, since no test uses it, but I will leave the support in.

File: tests/framework/utils.py (hunk 5 above)
Comment (pb8o) [reply tests/framework/utils.py:683]: yw :)!

File: tools/devctr/Dockerfile (hunk 11 above)
Comment (bchalios) [reply tools/devctr/Dockerfile:75]: awesome, ok. I 'll pick it up with my PR then.

File: tests/framework/microvm.py (hunk 10 above)
Comment (bchalios) [reply tests/framework/microvm.py:865]: cool! I didn't know that

File: tests/integration_tests/security/test_jail.py [hunk 12]
```diff
@@ -81,6 +66,7 @@ def test_empty_jailer_id(test_microvm_with_api):
     test_microvm.jailer = JailerContext(
         jailer_id="",
         exec_file=fc_binary,
+        netns=None,
```
Comment (bchalios) [inline tests/integration_tests/security/test_jail.py:69]: Here is a `netns=None`

File: tests/framework/jailer.py (hunk 6 above)
Comment (bchalios) [reply tests/framework/jailer.py]: I found an instance of `netns=None` added in this PR.

File: tests/framework/jailer.py (hunk 7 above)
Comment (pb8o) [reply tests/framework/jailer.py:105]: Ah yeah, that's a negative test. Actually we don't need to pass None explicitly now. That could be removed too.

File: tests/integration_tests/security/test_jail.py (hunk 12 above)
Comment (pb8o) [reply tests/integration_tests/security/test_jail.py:69]: Yep, I think that could be dropped, is not important for the test.

Comment (pb8o) [reply tests/integration_tests/security/test_jail.py]: fixed

**Review by bchalios — APPROVED** [review]

**Review by roypat — APPROVED** [review]

File: tests/integration_tests/security/test_jail.py (hunk 4 above)
Comment (pb8o) [reply tests/integration_tests/security/test_jail.py:94]: I think I will leave it for now, but I will think about it.
